In [1]:
%load_ext sql

In [2]:
!pip install SQLAlchemy==1.4.49

  Using cached SQLAlchemy-1.4.49-cp312-cp312-win_amd64.whl.metadata (10 kB)
Using cached SQLAlchemy-1.4.49-cp312-cp312-win_amd64.whl (1.6 MB)
  Attempting uninstall: SQLAlchemy
    Found existing installation: SQLAlchemy 2.0.41
    Uninstalling SQLAlchemy-2.0.41:
      Successfully uninstalled SQLAlchemy-2.0.41


  You can safely remove it manually.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
aext-project-filebrowser-server 4.20.0 requires sqlalchemy>=2.0.29, but you have sqlalchemy 1.4.49 which is incompatible.


In [3]:
!pip install pymysql

In [8]:
#Testing connection to DB

import pymysql
import pandas as pd

# Create connection
conn = pymysql.connect(
    host='localhost',
    user='root',
    password='Diagnostic_56',
    database='united_nations'
)

# Query
query = "SELECT * FROM united_nations.access_to_basic_services;"
df = pd.read_sql(query, conn)
df

C:\Users\User\AppData\Local\Temp\ipykernel_912\4061088695.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Region      Sub_region Country_name  Time_period  \
0     Central and Southern Asia    Central Asia   Kazakhstan         2015   
1     Central and Southern Asia    Central Asia   Kazakhstan         2016   
2     Central and Southern Asia    Central Asia   Kazakhstan         2017   
3     Central and Southern Asia    Central Asia   Kazakhstan         2018   
4     Central and Southern Asia    Central Asia   Kazakhstan         2019   
...                         ...             ...          ...          ...   
1043         Sub-Saharan Africa  Western Africa         Togo         2016   
1044         Sub-Saharan Africa  Western Africa         Togo         2017   
1045         Sub-Saharan Africa  Western Africa         Togo         2018   
1046         Sub-Saharan Africa  Western Africa         Togo         2019   
1047         Sub-Saharan Africa  Western Africa         Togo         2020   

      Pct_managed_drinking_water_services  Pct_managed_sanitation_services  \
0                                   94.67                            98.00   
1                                   94.67                            98.00   
2                                   95.00                            98.00   
3                                   95.00                            98.00   
4                                   95.00                            98.00   
...                                   ...                              ...   
1043                                67.33                            18.33   
1044                                68.33                            18.33   
1045                                69.33                            19.33   
1046                                70.33                            19.67   
1047                                70.67                            20.00   

      Est_population_in_millions  Est_gdp_in_billions  Land_area  \
0                      17.542806               184.39  2699700.0   
1                      17.794055               137.28  2699700.0   
2                      18.037776               166.81  2699700.0   
3                      18.276452               179.34  2699700.0   
4                      18.513673               181.67  2699700.0   
...                          ...                  ...        ...   
1043                    7.661354                 6.03    54390.0   
1044                    7.852795                 6.40    54390.0   
1045                    8.046679                 7.11    54390.0   
1046                    8.243094                 7.22    54390.0   
1047                    8.442580                 7.57    54390.0   

      Pct_unemployment  
0                 4.93  
1                 4.96  
2                 4.90  
3                 4.85  
4                 4.80  
...                ...  
1043               NaN  
1044              3.74  
1045               NaN  
1046               NaN  
1047               NaN  

[1048 rows x 10 columns]

In [11]:
#SELECT COLUMNS TO BE USED IN ANALYSIS

import pymysql
import pandas as pd

# Create connection
conn = pymysql.connect(
    host='localhost',
    user='root',
    password='Diagnostic_56',
    database='united_nations'
)

# Query
query = "SELECT Country_name, Time_period, Pct_managed_drinking_water_services FROM united_nations.Access_to_Basic_Services;"
df = pd.read_sql(query, conn)
df

C:\Users\User\AppData\Local\Temp\ipykernel_912\2148719394.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Country_name  Time_period  Pct_managed_drinking_water_services
0      Kazakhstan         2015                                94.67
1      Kazakhstan         2016                                94.67
2      Kazakhstan         2017                                95.00
3      Kazakhstan         2018                                95.00
4      Kazakhstan         2019                                95.00
...           ...          ...                                  ...
1043         Togo         2016                                67.33
1044         Togo         2017                                68.33
1045         Togo         2018                                69.33
1046         Togo         2019                                70.33
1047         Togo         2020                                70.67

[1048 rows x 3 columns]

In [15]:
# 1.Order countries based on their water access levels per year.

import pymysql
import pandas as pd

# Create connection
conn = pymysql.connect(
    host='localhost',
    user='root',
    password='Diagnostic_56',
    database='united_nations'
)

# Query
query = "SELECT Country_name, Time_period, Pct_managed_drinking_water_services, ROW_NUMBER() OVER( PARTITION BY Time_period ORDER BY Pct_managed_drinking_water_services ASC) AS Rank_of_water_services FROM united_nations.Access_to_Basic_Services;"
df = pd.read_sql(query, conn)
df


C:\Users\User\AppData\Local\Temp\ipykernel_912\1112341739.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Country_name  Time_period  \
0             Central African Republic         2015   
1     Democratic Republic of the Congo         2015   
2                          South Sudan         2015   
3                               Angola         2015   
4                              Somalia         2015   
...                                ...          ...   
1043                  French Polynesia         2020   
1044                           Tokelau         2020   
1045                            Tuvalu         2020   
1046                         Mauritius         2020   
1047                           Réunion         2020   

      Pct_managed_drinking_water_services  Rank_of_water_services  
0                                   44.00                       1  
1                                   45.33                       2  
2                                   46.33                       3  
3                                   50.33                       4  
4                                   50.67                       5  
...                                   ...                     ...  
1043                               100.00                     161  
1044                               100.00                     162  
1045                               100.00                     163  
1046                               100.00                     164  
1047                               100.00                     165  

[1048 rows x 4 columns]

In [17]:
# 2. Assess the rankings for countries with the same water level access in a given year.

import pymysql
import pandas as pd

# Create connection
conn = pymysql.connect(
    host='localhost',
    user='root',
    password='Diagnostic_56',
    database='united_nations'
)

# Query
query = "SELECT Country_name, Time_period, Pct_managed_drinking_water_services, ROW_NUMBER() OVER( PARTITION BY Time_period ORDER BY Pct_managed_drinking_water_services ASC) AS Rank_of_water_services FROM united_nations.Access_to_Basic_Services WHERE Pct_managed_drinking_water_services = 100;"
df = pd.read_sql(query, conn)
df


C:\Users\User\AppData\Local\Temp\ipykernel_912\3134949196.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Country_name  Time_period  Pct_managed_drinking_water_services  \
0                 Singapore         2015                                100.0   
1                   Bermuda         2015                                100.0   
2                 Greenland         2015                                100.0   
3    British Virgin Islands         2015                                100.0   
4                Guadeloupe         2015                                100.0   
..                      ...          ...                                  ...   
203        French Polynesia         2020                                100.0   
204                 Tokelau         2020                                100.0   
205                  Tuvalu         2020                                100.0   
206               Mauritius         2020                                100.0   
207                 Réunion         2020                                100.0   

     Rank_of_water_services  
0                         1  
1                         2  
2                         3  
3                         4  
4                         5  
..                      ...  
203                      35  
204                      36  
205                      37  
206                      38  
207                      39  

[208 rows x 4 columns]

In [24]:
# 3. Apply the RANK function instead.

import pymysql
import pandas as pd

# Create connection
conn = pymysql.connect(
    host='localhost',
    user='root',
    password='Diagnostic_56',
    database='united_nations'
)

# Query
query = "SELECT Country_name, Time_period, Pct_managed_drinking_water_services, RANK() OVER( PARTITION BY Time_period ORDER BY Pct_managed_drinking_water_services ASC) AS Rank_of_water_services FROM united_nations.Access_to_Basic_Services;"
df = pd.read_sql(query, conn)
df


C:\Users\User\AppData\Local\Temp\ipykernel_912\3174509531.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Country_name  Time_period  \
0             Central African Republic         2015   
1     Democratic Republic of the Congo         2015   
2                          South Sudan         2015   
3                               Angola         2015   
4                              Somalia         2015   
...                                ...          ...   
1043                  French Polynesia         2020   
1044                           Tokelau         2020   
1045                            Tuvalu         2020   
1046                         Mauritius         2020   
1047                           Réunion         2020   

      Pct_managed_drinking_water_services  Rank_of_water_services  
0                                   44.00                       1  
1                                   45.33                       2  
2                                   46.33                       3  
3                                   50.33                       4  
4                                   50.67                       5  
...                                   ...                     ...  
1043                               100.00                     127  
1044                               100.00                     127  
1045                               100.00                     127  
1046                               100.00                     127  
1047                               100.00                     127  

[1048 rows x 4 columns]

We can see that in 2015, the country with the lowest percentage of managed drinking water services was Central African Republic . Within the year 2020, we have countries that had the same percentage, such as French Polynesia and Tokelau, and as a result, they have the same rank, 127. 